fabric-client-tags-api-v1/src/client/java/net/fabricmc/fabric/api/tag/client/v1/ClientTags.java

```diff
@@ -175,4 +163,30 @@ private static <T> Optional<? extends Registry<T>> getRegistry(TagKey<T> tagKey)
 
 		return (Optional<? extends Registry<T>>) Registries.REGISTRIES.getOrEmpty(tagKey.registry().getValue());
 	}
+
+	@SuppressWarnings("unchecked")
+	private static <T> Optional<RegistryEntry<T>> getRegistryEntry(TagKey<T> tagKey, T entry) {
+		Optional<? extends Registry<?>> maybeRegistry = getRegistry(tagKey);
+
+		if (maybeRegistry.isEmpty() || !tagKey.isOf(maybeRegistry.get().getKey())) {
+			return Optional.empty();
+		}
+
+		Registry<T> registry = (Registry<T>) maybeRegistry.get();
+
+		Optional<RegistryKey<T>> maybeKey = registry.getKey(entry);
+
+		return maybeKey.map(registry::entryOf);
+	}
+
+	private static ClientTagsLoader.LoadedTag getOrCreatePartiallySyncedTag(TagKey<?> tagKey) {
+		ClientTagsLoader.LoadedTag loadedTag = LOCAL_TAG_HIERARCHY.get(tagKey);
+
+		if (loadedTag == null) {
+			loadedTag = ClientTagsLoader.loadTag(tagKey);
+			LOCAL_TAG_HIERARCHY.put(tagKey, loadedTag);
+		}
+
+		return loadedTag;
+	}
```

Comment:
Could maybe move these methods and `LOCAL_TAG_HIERARCHY` to the impl class? Generally I think its best to keep the public API classes free of a lot of logic, it does make it a little easier to se what is/isnt api. Im not too bothered either way though as this class already contained quite a bit of logic anyway 👍

Reply:
I'll leave it for now.